normandy/base/middleware.py

```diff
@@ -128,3 +136,109 @@ def process_request(self, request):
                 }
             )
         return response
+
+
+def default_username(email):
+    """Generate username for the Django user.
+
+    :arg str/unicode email: the email address to use to generate a username
+
+    :returns: str/unicode
+
+    """
+    # bluntly stolen from django-browserid
+    # store the username as a base64 encoded sha224 of the email address
+    # this protects against data leakage because usernames are often
+    # treated as public identifiers (so we can't use the email address).
+    return smart_text(base64.urlsafe_b64encode(
+        hashlib.sha1(force_bytes(email)).digest()
+    ).rstrip(b'='))
+
+
+class OIDCAccessTokenAuthenticationMiddleware:
+
+    def __init__(self, get_response):
+        self.get_response = get_response
+        # XXX Here we *could* do something like:
+        # if not settings.OIDC_ACCESS_TOKEN_ENABLED:
+        #    from django.core.exceptions import MiddlewareNotUsed
+        #    raise MiddlewareNotUsed
+        self.oidc_user_endpoint = settings.OIDC_USER_ENDPOINT
```

Comment:
Just a nice feeling that once the middleware is instanciated, it's complete and doesn't need to re-read other sources.

Reply:
It also means that testing is harder, since a test can't easily change a global setting and then test the results. In general, I think that storing the values from settings mostly leads to trouble.

Reply:
No, the middleware is instanciated for every test. So if you change `settings.OIDC_USER_ENDPOINT` in just one test it changes for that request (if the test makes a request)

Reply:
I had to check, out of curiosity. If you run django's runserver, it creates and stores instances of the middleware classes between requests.

Reply:
Accessing settings isn't an expensive operation. Storing settings off to the side creates different views into what the settings actually are. This is a minor issue here, but code that relies on settings not changing is fragile and has the potential to become a maintenance issue later. I'd prefer not to do it, even if it would work in the short term.